### DISCUSSION TITLE
[Libcurl 7.80.0 + Openssl-3.0.7]: Getting (35) Insufficient randomness error during SSL connect via libcurl

### DISCUSSION THREAD

**Discussion by bodachaitanya** [curl/curl#11022] [Q&A] [answered]
Getting SSL Connect error (35) with "Insufficient randomness" when using libcurl-7.80.0 with Openssl-3.0.7 for SSL connection to server via libcurl related API's. Here are the details:
- Openssl-3.0.7 built with configure options: `shared, zlib`
- Libcurl-7.80.0 built with configure options: `--with-openssl, --enable-openssl-auto-load-config, --without-libssh2, --disable-ldap, CFLAGS="-O2 -g -w -DCAVIUM_SSL -DMC2 -g -O", CXX=g++, CC=cc`
- Only above libcurl and openssl versions work for getting successful libcurl build. Although, I have not seen mentioned about versions compatibility anywhere in the web, with lower versions of libcurl than 7.80.0, builds with Openssl-3.x is not building, so it looks like only these versions (and higher) for both software supports.
- Linux kernel version: `4.17.00`
- As per checking, PR https://github.com/curl/curl/pull/7856 introduced in libcurl-7.80.0 is the change which is causing this error.
- Multiple options are tried in the process of fixing the error, but was not helpful. Details of options tried:
  - Building with older than libcurl-7.80.0 with openssl-3.0 -- Does not work
  - Building with older than libcurl-7.80.0 with openssl-3.0 -- Does not work
  - Kernel entropy improvement patch - Did not work
    - Backported kernel entropy improvement from 5.15.16 kernel to 4.17.00 kernel
  - Random seed with OS (although OS is default, but tried with explicit build to seed randomness from OS) -- Did not work
  - Random seed with CPU (rdcpu) -- recommended for older kernels than 5.18 -- Did not work
  - As per known issues and discussions on the same issues (like :
    - Tried provider (default) setting explicitly for openssl -- Did not work
  - Updated libcurl code to add randomness explicitly -- Did not work
- The issue is happening because RAND_status() in libcurl (in lib/vtls/openssl.c) is always returning FALSE, irrespective of randomness is sufficient or not. Randomness on the node seems to be ok, as per checking with test code provided in openssl (test directory) and also using below code snippet:
```
#include <openssl/rand.h>
#include <stdio.h>
int main(int argc, char **argv)
{ 
	int r1 = RAND_poll(); 
	printf("RAND_poll() said %d\n", r1); 
	unsigned char byte; 
	int r2 = RAND_bytes(&byte, 1); 
	printf("RAND_bytes() said %d\n", r2); 
	int r3 = RAND_status(); 
	printf("RAND_status() said %d\n", r3); 
	if (r3 == 1) { 
		puts("RAND_status is okay!"); 
	} else { 
	puts("RAND_status said we aren't seeded?"); 
	} 
	return 0; 
}

Output:
--------
RAND_poll() said 1
RAND_bytes() said 1
RAND_status() said 1
RAND_status is okay!
```
**Note**: This issue is seen when trying to migrate both libcurl and openssl versions to support openssl-3.x from the current libcurl-7.74.0 and openssl-1.1.1.

Could you please help with what can be the issue here. Looks like it is not actual issue with randomness insufficiency, but libcurl itself is blocking in its code checking RAND_status() output. Not sure why RAND_status() in libcurl is throwing always FALSE, probably randomness seeding inside libcurl is not sufficient which I am not able to figure out from libcurl code. As I understand, if backend (like openssl) supports random number generation, then libcurl does no random number generation by itself. Then, why does it fail at this point. Any pointers and help will be helpful.

**Comment by bagder** [comment]
> RAND_status() in libcurl is throwing always FALSE

This points to it being an OpenSSL issue I think. OpenSSL3 seeds and inits its own random and it should then not ever return error for this function if everything is fine. curl refuses to continue because OpenSSL says it's random status is not okay. I don't see what curl can do differently about this. You need to investigate why OpenSSL tells us this.

**Reply by bodachaitanya** [reply]
From the sample code which I shared in the query, RAND_status() shows fine (returns TRUE/1), even without RAND_poll() and RAND_bytes() call as well. It is only with libcurl, with this PR https://github.com/curl/curl/pull/7856, I am seeing RAND_status() not returning TRUE on the same node. Any comments on this?

**Reply by bagder** [reply]
I am not an OpenSSL expert, I don't know why it returns error in this case but not in other cases.

(Of course curl did not return that error before that change was merged that introduced it properly bailing out on this error.)

**Comment by bodachaitanya** [comment]
Hi @bagder ,

I tried checking further by using sample programs with easy handle, multi handle and multi socket options on the same node with above configuration, and seems like multi socket has an issue with TLS connection establishment with the server. Please find the details below:

### Using curl easy handle
#### Test program
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

#define URL "https://example.org"

// Callback function for SSL certificate verification
int ssl_callback(CURL *curl, void *sslctx, void *userptr)
{
    // Load CA certificate file
    CURLcode result = CURLE_OK;
    const char *cacert_file = "cacert.crt";

    if (curl_easy_setopt(curl, CURLOPT_CAINFO, cacert_file) != CURLE_OK)
    {
        printf("Failed to load CA certificate file.\n");
        result = CURLE_SSL_CERTPROBLEM;
    }
    return result;
}

int main()
{
    CURL *easy_handle;
    CURLcode result;

    // Initialize libcurl
    curl_global_init(CURL_GLOBAL_ALL);

    // Create easy handle
    easy_handle = curl_easy_init();
    curl_easy_setopt(easy_handle, CURLOPT_URL, URL);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYPEER, 1);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYHOST, 2);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_CTX_FUNCTION, ssl_callback);

    // Perform SSL connection
    result = curl_easy_perform(easy_handle);
    if (result != CURLE_OK)
    {
        printf("Failed to perform SSL connection: %s\n", curl_easy_strerror(result));
    }

    // Cleanup
    curl_easy_cleanup(easy_handle);
    curl_global_cleanup();

    return 0;
}
```
#### Result
- TLS connection worked and able to connect to both nSA and also external URL (example.org).
- Wireshark packet trace showing TLS connection establishment:
![image](https://user-images.githubusercontent.com/16359474/234630959-aa79abb7-0876-445e-a59f-746940248365.png)


### Using curl multi handle
#### Test program
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

#define URL "https://example.org"
#define MAX_WAIT_MSECS 30 * 1000 /* Wait max. 30 seconds */

// Callback function for SSL certificate verification
int ssl_callback(CURL *curl, void *sslctx, void *userptr)
{
    // Load CA certificate file
    CURLcode result = CURLE_OK;
    const char *cacert_file = "cacert.crt";

    if (curl_easy_setopt(curl, CURLOPT_CAINFO, cacert_file) != CURLE_OK)
    {
        printf("Failed to load CA certificate file.\n");
        result = CURLE_SSL_CERTPROBLEM;
    }
    return result;
}

int main()
{
    CURL *easy_handle;
    CURLM *multi_handle;
    CURLcode result;
    int still_running;
    int num_handles = 0;

    // Initialize libcurl
    curl_global_init(CURL_GLOBAL_ALL);

    // Create easy handle and set options
    easy_handle = curl_easy_init();
    curl_easy_setopt(easy_handle, CURLOPT_URL, URL);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYPEER, 1);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYHOST, 2);
    curl_easy_setopt(easy_handle, CURLOPT_SSL_CTX_FUNCTION, ssl_callback);

    // Create multi handle and add easy handle to it
    multi_handle = curl_multi_init();
    curl_multi_add_handle(multi_handle, easy_handle);
    num_handles++;

    // Perform SSL connection(s)
    do {
        result = curl_multi_perform(multi_handle, &still_running);
        if (result != CURLM_OK)
        {
            printf("Failed to perform SSL connection: %s\n", curl_easy_strerror(result));
            break;
        }

        // Wait for activity, timeout or until interrupted
        CURLMcode mc = curl_multi_poll(multi_handle, NULL, 0, MAX_WAIT_MSECS, NULL);
        if (mc != CURLM_OK)
        {
            printf("curl_multi_poll() failed: %s\n", curl_multi_strerror(mc));
            break;
        }

    } while (still_running);

    // Remove easy handle from multi handle and cleanup
    curl_multi_remove_handle(multi_handle, easy_handle);
    curl_easy_cleanup(easy_handle);
    curl_multi_cleanup(multi_handle);
    curl_global_cleanup();

    return 0;
}
```
#### Result
- TLS connection worked and able to connect to both nSA and also external URL (example.org).
- Wireshark packet trace showing TLS connection establishment:
![image](https://user-images.githubusercontent.com/16359474/234631871-0cb7570a-f9f7-4a11-9832-375ec09d1aec.png)

### Using curl multi socket action:
#### Test program
```
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <curl/curl.h>

#define URL "https://example.org"

// Callback function for SSL certificate verification
int ssl_callback(CURL *curl, void *sslctx, void *userptr)
{
  // Load CA certificate file
  CURLcode result = CURLE_OK;
  const char *cacert_file = "cacert.crt";

  if (curl_easy_setopt(curl, CURLOPT_CAINFO, cacert_file) != CURLE_OK) {
    printf("Failed to load CA certificate file.\n");
    result = CURLE_SSL_CERTPROBLEM;
  }
  return result;
}

int main()
{
  CURLM *multi_handle;
  CURL *easy_handle;
  int still_running;
  int i;

  // Initialize libcurl
  curl_global_init(CURL_GLOBAL_ALL);

  // Create multi handle
  multi_handle = curl_multi_init();

  // Create easy handle
  easy_handle = curl_easy_init();
  curl_easy_setopt(easy_handle, CURLOPT_URL, URL);
  curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYPEER, 1);
  curl_easy_setopt(easy_handle, CURLOPT_SSL_VERIFYHOST, 2);
  curl_easy_setopt(easy_handle, CURLOPT_SSL_CTX_FUNCTION, ssl_callback);

  // Add easy handle to multi handle
  curl_multi_add_handle(multi_handle, easy_handle);

  // Execute multi handle until all transfers are done
  do
  {
    curl_multi_socket_action(multi_handle, CURL_SOCKET_TIMEOUT, 0, &still_running);
  } while (still_running);

  // Cleanup
  curl_multi_remove_handle(multi_handle, easy_handle);
  curl_easy_cleanup(easy_handle);
  curl_multi_cleanup(multi_handle);
  curl_global_cleanup();

  return 0;
}
```
#### Result
- TLS connection failed to both nSA and also external URL (example.org).
- Wireshark packet trace showed no TLS packets:
![image](https://user-images.githubusercontent.com/16359474/234633228-2387f47c-dbb8-4faa-a3c1-69a4732817b3.png)

### Observation
- It looks like curl_multi_socket_action() has an issue with libcurl-7.80.0.  
- With curl_easy_perform() and curl_multi_perform(), TLS connection was established fine, and only getting failed with curl_multi_socket_action(), which we are using.

Could you please check and confirm if there is an issue with curl multi socket option way of connecting to server. Please let me know if any further details are required. Thanks

**Reply by bodachaitanya** [reply]
Hi @bagder , Could you please help here. Thanks

**Reply by bagder** [reply]
What makes you think calling `curl_multi_socket_action` with nothing but `CURL_SOCKET_TIMEOUT` is going to do the right thing for you? You never tell curl there is data to read (nor that it can write).

I don't think your test code is correct so there is no conclusion to draw from that.

**Reply by bodachaitanya** [reply]
The curl documentation about multi-socket in section [How to start everything](https://everything.curl.dev/libcurl/drive/multi-socket#how-to-start-everything) has this information:
```
To kick it all off, you tell libcurl it timed out (because all easy handles start out with a short timeout) 
which will make libcurl call the callbacks to set things up and from then on you can just let 
your event system drive:

/* all easy handles and callbacks are setup */

curl_multi_socket_action(multi, CURL_SOCKET_TIMEOUT, 0, &running);

/* now the callbacks should have been called and we have sockets to wait for
   and possibly a timeout, too. Make the event system do its magic */

event_base_dispatch(event_base); /* libevent2 has this API */

/* at this point we have exited the event loop */
```

**Reply by bagder** [reply]
It says that's how you start the engine, yes. But it does not say that you can then keep calling it saying it timed out and that will work and complete all the transfers. That's just not how that API is meant to be used nor designed to work.

**Reply by jay** [reply]
Yup. The code does not make sense, it's missing an event system. I suggest use regular multi instead, it's a lot easier. Also, setting CURLOPT_CAINFO option in your SSL_CTX callback does not make sense (though manual import and caching CAs through that callback may make some sense in a very advanced case).

**Reply by bodachaitanya** [reply]
Hi @jay - Currently we have curl_multi_socket_action() with event based mechanism (using epoll) implementation for connections to server in our product, and I was trying to simulate the same using test code to show about the issue we are encountering. I am trying to update test code with required FD's of sockets (i.e., active sockets)  and required data before multi socket action call. Any pointers from you would help too. Thanks!

Note: The same implementation exists since long (nearly 7+ years) and it seems to be not working as expected with curl upgraded to 7.80.0 along with Openssl-3.0.7. So, reverting back to multi_handle way is not an option, and want to check why multi_socket_action is failing and ultimately throwing 'Insufficient randomness" error during SSL connect.

**Reply by bagder** [reply]
It seems unlikely that "Insufficient randomness" happens differently depending on how you drive libcurl.

**Reply by bodachaitanya** [reply]
With the same multi_socket_action test code above, when I try to connect to an internal server, this is what it shows:
- TLS client hello is initiated:
![image](https://user-images.githubusercontent.com/16359474/235078918-37ebf451-cc8e-484a-9a78-0c73c70cd194.png)
- No server response received and fails with this alert:
![image](https://user-images.githubusercontent.com/16359474/235079411-e482db66-d2ab-4db1-965d-af1a982856e8.png)
- It is during TLS client hello stage (as in 1st step), random number is generated, and looks like curl with multi_socket_action is inherently checking its randomness strength and failing always with "Insufficient randomness" error.

**Comment by anthonyryan1** [comment]
I think I'm seeing this same issue immediately after upgrading from OpenSSL 1.1.1 to OpenSSL 3.0.

I'm using a higher level language, so I didn't get the SSL error, just a "returned no response".

I can confirm going back to OpenSSL 1.1.1 resolved the issue. In my case, our application exclusively uses multi handle (even when we're only dispatching one request). If I dispatch a few requests, everything works. But it fails almost every time when I send 120 requests, each to a unique host, and dispatch them all at the same time.

Is it possible that with this many requests launching at the exact same time, randomness is actually temporarily running out? It seems feasible because I've got 120 consumers starting at the same time.

I will keep experimenting with this for the moment.

**Comment by anthonyryan1** [comment]
I have confirmed the issue also is fixed when curl is built against OpenSSL 3.1.1. This seems to be a bug exclusive to OpenSSL 3.0.x.

**Comment by swati-ISW** [comment]
Hi @anthonyryan1 ,
With what curl version did you test this out. I can still see the same issue with 7.80.0+openssl 3.1.1

**Reply by anthonyryan1** [reply, marked as answer]
I'm using `libcurl/8.0.1` +  `OpenSSL/3.1.1` which resolved the issues I experienced.
